### PR TITLE
Add abilitiy to subscribe to changes of ViewModel properties

### PR DESCRIPTION
**Pull request by brooklynDev**
Often times it is useful to react in code to changes on the ViewModel properties within the ViewModel itself. One way to do that is to add code within the setter of that property itself:

```
public class MyViewModel : ViewModelBase
{

   private string _myValue;
   public string MyValue
   {
      get => _myValue;
      set 
      {
         Set(ref _myValue, value);
         DoSomeWork();
      }
}
```

This however is a bit clunky and hard to follow. I think it would be nicer to have an explicit way to Subscribe to changes. This PR introduces the following:

```
public class MyViewModel : ViewModelBase
{
   public MyViewModel()
   {
       Subscribe(() => MyValue, newValue => DoSomeWork());
   }
   private string _myValue;
   public string MyValue
   {
      get => _myValue;
      set => Set(ref _myValue, value);
   }
}
```

### CODE REVIEW

File: src/MvvmBlazor/ViewModel/ViewModelBase.cs [hunk 1]
```diff
@@ -17,6 +22,14 @@ protected bool Set<T>(ref T field, T value, [CallerMemberName] string? propertyN
             {
                 field = value;
                 OnPropertyChanged(propertyName!);
+                if (!_subscriptions.ContainsKey(propertyName))
```
Comment (klemmchr) [inline src/MvvmBlazor/ViewModel/ViewModelBase.cs:25]: Could you fix the warnings for possible null references?

Comment (klemmchr) [inline src/MvvmBlazor/ViewModel/ViewModelBase.cs:25]: Could you fix the warnings for possible null references?

**Review by klemmchr — CHANGES_REQUESTED** [review]

File: src/MvvmBlazor/ViewModel/ViewModelBase.cs (hunk 1 above)
Comment (brooklynDev) [reply src/MvvmBlazor/ViewModel/ViewModelBase.cs]: Done!

File: src/MvvmBlazor/ViewModel/ViewModelBase.cs [hunk 2]
```diff
@@ -28,6 +41,24 @@ public virtual void OnPropertyChanged(string propertyName)
             PropertyChanged?.Invoke(this, new PropertyChangedEventArgs(propertyName));
         }
 
+        protected void Subscribe<T>(Expression<Func<T>>? expression, Action<T> action)
+        {
+            var member = expression?.Body as MemberExpression;
+            var propInfo = member?.Member as PropertyInfo;
+            if (propInfo == null)
+            {
```
Comment (klemmchr) [inline src/MvvmBlazor/ViewModel/ViewModelBase.cs:58]: I'm not sure whether this should just return in this case. Throwing a proper error should be better, especially if the property does not exist.

**Review by klemmchr — CHANGES_REQUESTED** [review]

**Comment by brooklynDev** [comment]
Sorry for the confusion - but I just updated the PR to add the ability to subscribe using either a synchronous function (`Action<T>`) or an async function (`Func<T, Task>`). So there's now a `Subscribe` and a `SubscribeAsync` function. Let me know what you think and I'll make changes as needed.

File: src/MvvmBlazor/ViewModel/ViewModelBase.cs (hunk 2 above)
Comment (brooklynDev) [reply src/MvvmBlazor/ViewModel/ViewModelBase.cs:58]: That makes sense. Which exception would you propose I throw?

Comment (klemmchr) [reply src/MvvmBlazor/ViewModel/ViewModelBase.cs:58]: There already exists a `BindingException` that could be used here.

Comment (brooklynDev) [reply src/MvvmBlazor/ViewModel/ViewModelBase.cs:57]: Updated

**Comment by klemmchr** [comment]
It's missing verification of the property expression, however, I would just leave it like this for now. In order to be able to check for the right property we would need some caching so performance isn't affected.
